### PR TITLE
refactor: staking factory to check mech and requester staking instances

### PR DESCRIPTION
**Pull request by kupermind**
- Staking factory to check mech and requester staking instances.

### CODE REVIEW

File: contracts/AgentMech.sol [hunk 1]
```diff
@@ -174,37 +180,6 @@ contract AgentMech is ERC721Mech {
         }
     }
 
-    /// @dev Performs actions before the delivery of a request.
-    /// @param data Self-descriptive opaque data-blob.
-    /// @return requestData Data for the request processing.
-    function _preDeliver(address, uint256, bytes memory data) internal virtual returns (bytes memory requestData) {
```
Comment (kupermind) [inline contracts/AgentMech.sol:180]: Not deleting, just moving around to have a better contract visibility.

File: contracts/MechMarketplace.sol [hunk 2]
```diff
@@ -246,34 +251,11 @@ contract MechMarketplace {
         );
     }
 
-    /// @dev Checks agent mech for contract validity.
-    /// @param mech Agent mech address.
-    /// @param mechServiceId Mech operator service Id.
-    function checkMech(address mech, uint256 mechServiceId) public view {
```
Comment (kupermind) [inline contracts/MechMarketplace.sol:252]: These are changed to ones below

File: contracts/MechMarketplace.sol [hunk 3]
```diff
@@ -311,20 +294,10 @@ contract MechMarketplace {
...
-        }
+        checkMech(priorityMech, priorityMechStakingInstance, priorityMechServiceId);
 
-        // Get the staked service info
-        IStaking.ServiceInfo memory serviceInfo = IStaking(requesterStakingInstance).getServiceInfo(requesterServiceId);
-        // Check staked service multisig
-        if (serviceInfo.multisig != msg.sender) {
-            revert OwnerOnly(msg.sender, serviceInfo.multisig);
-        }
+        // Check requester
```
Comment (kupermind) [inline contracts/MechMarketplace.sol:299]: Mechs and requester services are verified in these functions

File: contracts/MechMarketplace.sol [hunk 4]
```diff
@@ -446,6 +435,54 @@ contract MechMarketplace {
         ));
     }
 
+    /// @dev Checks for staking instance contract validity.
+    /// @param stakingInstance Staking instance address.
+    /// @param serviceId Service Id.
+    function checkStakingInstance(address stakingInstance, uint256 serviceId) public view {
```
Comment (kupermind) [inline contracts/MechMarketplace.sol:441]: This function verifies that the staking contract of either requester / mech is valid and that the corresponding service is staked.

File: contracts/MechMarketplace.sol [hunk 5]
```diff
@@ -446,6 +435,54 @@ contract MechMarketplace {
...
+        if (state != IStaking.StakingState.Staked) {
+            revert ServiceNotStaked(stakingInstance, serviceId);
+        }
+    }
+
+    /// @dev Checks for mech validity.
+    /// @dev mech Agent mech contract address.
+    /// @param mechStakingInstance Agent mech staking instance address.
+    /// @param mechServiceId Agent mech service Id.
+    function checkMech(address mech, address mechStakingInstance, uint256 mechServiceId) public view {
```
Comment (kupermind) [inline contracts/MechMarketplace.sol:458]: This function then additionally checks that the mech is correct

File: contracts/MechMarketplace.sol [hunk 6]
```diff
@@ -446,6 +435,54 @@ contract MechMarketplace {
...
+        if (!IMech(mech).isOperator(serviceInfo.multisig)) {
+            revert UnauthorizedAccount(mech);
+        }
+    }
+
+    /// @dev Checks for requester validity.
+    /// @dev requester Requester contract address.
+    /// @param requesterStakingInstance Requester staking instance address.
+    /// @param requesterServiceId Requester service Id.
+    function checkRequester(address requester, address requesterStakingInstance, uint256 requesterServiceId) public view {
```
Comment (kupermind) [inline contracts/MechMarketplace.sol:474]: This function is additionally checkin that the requester is correct.

**Review by DavidMinarsch — COMMENTED** [review]
nice!

**Pull request merged**